src/esp/assets/PTexMeshData.cpp

```diff
@@ -602,35 +616,56 @@ void PTexMeshData::uploadBuffersToGPU(bool forceReload) {
                                   currentMesh->abo);
     currentMesh->abo.setData(adjFaces[iMesh],
                              Magnum::GL::BufferUsage::StaticDraw);
+
+    // experiment code (may not work):
     currentMesh->mesh.setPrimitive(Magnum::GL::MeshPrimitive::LinesAdjacency)
         .setCount(currentMesh->ibo.size() / 2)
         .addVertexBuffer(currentMesh->vbo, 0, gfx::PTexMeshShader::Position{})
         .setIndexBuffer(currentMesh->ibo, 0,
                         Magnum::GL::MeshIndexType::UnsignedInt);
   }
 
+  // load atlas data and upload them to GPU
+
+  std::cout << "loading atlas textures: " << std::endl;
   for (size_t iMesh = 0; iMesh < renderingBuffers_.size(); ++iMesh) {
-    const std::string rgbFile =
-        atlasFolder_ + "/" + std::to_string(iMesh) + "-color-ptex.rgb";
-    if (!io::exists(rgbFile)) {
-      ASSERT(false, "Can't find " + rgbFile);
-    }
+    const std::string rgbFile = Corrade::Utility::Directory::join(
+        atlasFolder_, std::to_string(iMesh) + "-color-ptex.rgb");
+
+    ASSERT(io::exists(rgbFile), Error : Cannot find the rgb file);
+
     std::cout << "\rLoading atlas " << iMesh + 1 << "/"
-              << renderingBuffers_.size() << "... ";
+              << renderingBuffers_.size() << " from " << rgbFile << "... ";
     std::cout.flush();
 
-    Cr::Containers::Array<const char, Cr::Utility::Directory::MapDeleter> data =
-        Cr::Utility::Directory::mapRead(rgbFile);
-    const int dim = static_cast<int>(std::sqrt(data.size() / 3));  // square
+    const size_t numBytes = io::fileSize(rgbFile);
+    const int dim = static_cast<int>(std::sqrt(numBytes / 3));  // square
+
+    // Open file
+    int fd = open(std::string(rgbFile).c_str(), O_RDONLY, 0);
+    // MAP_POPULATE does not work on mac. It is to reduce the penalty of page
+    // faults. Code should be OK without it. void* mmappedData = mmap(NULL,
+    // numBytes, PROT_READ, MAP_PRIVATE | MAP_POPULATE, fd, 0);
+    void* mmappedData = mmap(NULL, numBytes, PROT_READ, MAP_PRIVATE, fd, 0);
+
+    Corrade::Containers::ArrayView<unsigned char> data(
+        (unsigned char*)(mmappedData), numBytes);
+
+    // atlas
+    // the size of each image is dim x dim x 3 (RGB), which equals to numBytes
     Magnum::ImageView2D image(Magnum::PixelFormat::RGB8UI, {dim, dim}, data);
     renderingBuffers_[iMesh]
         ->tex.setWrapping(Magnum::GL::SamplerWrapping::ClampToEdge)
         .setMagnificationFilter(Magnum::GL::SamplerFilter::Linear)
         .setMinificationFilter(Magnum::GL::SamplerFilter::Linear)
-        // .setStorage(1, GL::TextureFormat::RGB8UI, image.size())
+        // .setStorage(1, Magnum::GL::TextureFormat::RGB8UI, image.size())
```

Comment:
@mosra : 
The code in the master version commented it out.
Do I need to set the storage here?

Reply:
It was commited this way by @msavva in October 2018, I don't know what was the original reason :) Generally, calling `setSubImage()` without `setStorage()` or `setImage()` being called first will result in a GL error and nothing being done, so I guess that's what is happening here? (Unless the texture storage is being set up somewhere else, but I don't see any such code anywhere else.)

If you are on linux you can run the viewer with `--magnum-gpu-validation on` and it'll tell you in the console output about any GL errors. (Mac doesn't implement GL debug output, so there you'd need to test for `GL::Renderer::error()` manually by hand.)

Reply:
I am developing everything on the Mac. I guess I will have to require a linux box. :)

So here, if I understand you correctly, I should actively call setStorage(). Correct?

Reply:
Yes, or change the below to `setImage()` ... but `setStorage()` is the preferred thing to do.